compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -775,6 +788,10 @@ impl<'tcx> ValidateBoundVars<'tcx> {
 impl<'tcx> TypeVisitor<'tcx> for ValidateBoundVars<'tcx> {
     type BreakTy = ();
 
+    fn tcx_for_anon_const_substs(&self) -> TyCtxt<'tcx> {
```

Comment:
Why *not* pass in a tcx everywhere? (and name this `fn tcx`?)

Reply:
Some visitors **must not** look into the default substs, as that would cause query cycles. For others a `TyCtxt` isn't
readily available.

The second issue can be fixed by adding the `TyCtxt` as an additional argument to a bunch of places, but I would like
to avoid doing that in this PR (and probably in general tbh).

The first issue could also be solved by overwriting `visit_unevaluated_const` for the affected visitors, but I prefer doing it this way for aesthetic reasons :laughing:

Reply:
I also unfortunately started with `fn tcx_for_anon_const_substs(&self) -> TyCtxt<'tcx>` and changed it to
`fn tcx_for_anon_const_substs(&self) -> Option<TyCtxt<'tcx>>` which isn't too great tbh :sweat_smile: